lib/containers/k8s.pm

```diff
@@ -17,77 +17,115 @@ use testapi;
+    assert_script_run('uname -a');
+    assert_script_run('test -e /etc/rancher/k3s/k3s.yaml');
+    assert_script_run('k3s check-config');
+    validate_script_output('k3s kubectl config get-clusters', qr/default/);
+    validate_script_output('k3s kubectl config get-users', qr/default/);
+    validate_script_output('k3s kubectl config get-contexts --no-headers=true -o name', qr/default/);
+    assert_script_run('k3s kubectl config view --raw');
+    validate_script_output_retry("k3s kubectl get nodes", qr/ Ready.*control-plane,master /, retry => 6, delay => 15, timeout => 90);
+    validate_script_output_retry("k3s kubectl get namespaces", qr/default.*Active/, timeout => 120, delay => 60, retry => 3);
+    validate_script_output_retry('k3s kubectl get events -A', qr/Started container coredns/, retry => 6, delay => 30, timeout => 90);
+
+    # the default service account should be ready by now
+    assert_script_run("k3s kubectl get serviceaccount default -o name");
+    # expect that k3s api to be ready and is accessible
+    record_info("k3s api resources", script_output("k3s kubectl api-resources"));
+    assert_script_run("k3s kubectl auth can-i 'create' 'pods'");
+    assert_script_run("k3s kubectl auth can-i 'create' 'deployments'");
+}
+
+sub ensure_k3s_start {
+    systemctl('start k3s');
+    systemctl('is-active k3s');
+}
+
+sub set_custom_registry {
+    return unless get_var('REGISTRY');
+
+    script_run("mkdir -p /etc/rancher/k3s");
+    my $registry = registry_url();
+    assert_script_run "curl " . data_url('containers/registries.yaml') . " -o /etc/rancher/k3s/registries.yaml";
+    file_content_replace("/etc/rancher/k3s/registries.yaml", REGISTRY => $registry);
+}
+
+sub setup_and_check_k3s {
+    set_custom_registry;
+    ensure_k3s_start;
+    script_run("rm -f ~/.kube");
+    script_run('mkdir -p ~/.kube/');
+    assert_script_run('ln -s /etc/rancher/k3s/k3s.yaml ~/.kube/config');
+    check_k3s;
+}
+
 =head2 install_k3s
-Installs k3s, checks the instalation and prepare the kube/config
+Deploy k3s using k3s-install script that is either pulled from upstream or distro
 =cut
 
 sub install_k3s {
-    if (get_var("DISTRI") eq "alp") {
-        assert_script_run('/usr/bin/k3s-install');
-    }
-    else {
-        if (is_microos()) {
-            trup_call('pkg install k3s-selinux');
-            check_reboot_changes;
-        }
-        zypper_call('in apparmor-parser') if (is_sle('>=15-sp1'));
-  # Apply additional options. For more information see https://rancher.com/docs/k3s/latest/en/installation/install-options/#options-for-installation-with-script
-        my $k3s_version = get_var("CONTAINERS_K3S_VERSION");
-        if ($k3s_version) {
-            record_info('k3s forced version', $k3s_version);
-            assert_script_run("export INSTALL_K3S_VERSION=$k3s_version");
+    # k3s might be already installed by default
+    return if (script_run('which k3s') == 0);
```

Comment:
If we're doing this, then I suggest to rename the subroutine to `install_k3s_when_needed`

Reply:
Agreed to leave it like this for the time being (PR is already big enough)